### PR TITLE
add distributed versions of a few common functions

### LINKED ISSUE #8: sum distributed array  on cluster very very slow
following code very very slow on sum(DA), it takes ten minutes and still no result,so I interrupted it use ctrl+c.
this shouldn't  happens if DistributedArrays  operations are performed over the localpart chunks and then aggregated. maybe there are some issues? 
~~~julia
 x=Array(ASCIIString,6)
 x[1]="arch01"
 x[2]="arch02"
 x[3]="arch03"
 x[4]="arch04"
 x[5]="arch05"
 x[6]="arch06"
addprocs(x)
@everywhere using DistributedArrays
A = fill(1.1, (1000,1000));
sum(A)
DA = distribute(A);
sum(DA)
sum(A) == sum(DA)

### PR DESCRIPTION
**Pull request by amitmurthy**
closes #8

### CODE REVIEW

**Comment by amitmurthy** [comment]
cc: @jakebolewski

File: src/DistributedArrays.jl [hunk 1]
```diff
@@ -487,4 +487,21 @@ Base.scale!(A::DArray, x::Number) = begin
     return A
 end
 
+Base.sum(d::DArray) = reduce(Base.AddFun(), d)
+Base.maximum(d::DArray) = reduce(Base.MaxFun(), d)
+Base.minimum(d::DArray) = reduce(Base.MinFun(), d)
+Base.maxabs(d::DArray) = maximum([fetch(r) for r in [@spawnat p maxabs(localpart(d)) for p in procs(d)]])
```
Comment (andreasnoack) [inline src/DistributedArrays.jl]: `mapreduce` is also implemented for `DistributedArryas` so these could be written as e.g. `mapreduce(Base.AbsFun(), Base.MaxFun(),d)`

Comment (amitmurthy) [inline src/DistributedArrays.jl]: I tried it and it is very slow.

```
julia> typeof(d)
DistributedArrays.DArray{Int64,2,Array{Int64,2}}

julia> size(d)
(100,1000)

julia> @time mapreduce(Base.AbsFun(), Base.MaxFun(),d)
elapsed time: 8.103819938 seconds (730 MB allocated, 0.91% gc time in 33 pauses with 0 full sweep)
50

julia> @time maximum([fetch(r) for r in [@spawnat p maxabs(localpart(d)) for p in procs(d)]])
elapsed time: 0.006926139 seconds (89 kB allocated)
50
```

Comment (andreasnoack) [inline src/DistributedArrays.jl]: Ah. [This](https://github.com/JuliaParallel/DistributedArrays.jl/blob/1eef9381917858271088e8d21639fec09e48aa02/src/DistributedArrays.jl#L426) signature is too restrictive. It excludes the functors so it falls back on the method in Base. If the `Functions` is removed, I think it should work.

**Comment by jakebolewski** [comment]
lgtm, I agree with @andreasnoack that these are more natural to express with mapreduce.

File: src/DistributedArrays.jl (hunk 1 above)
Comment (amitmurthy) [inline src/DistributedArrays.jl]: Changing the `mapreduce` signature causes an ambiguity warning:

```
Warning: New definition 
    mapreduce(Any,Any,DistributedArrays.DArray{T,N,A}) at /home/amitm/.julia/v0.4/DistributedArrays/src/DistributedArrays.jl:415
is ambiguous with: 
    mapreduce(Any,Function,AbstractArray{T,N}) at reduce.jl:156.
```

I suggest leaving it as is.

File: src/DistributedArrays.jl [hunk 2]
```diff
@@ -487,4 +487,21 @@ Base.scale!(A::DArray, x::Number) = begin
...
+Base.minabs(d::DArray) = minimum([fetch(r) for r in [@spawnat p minabs(localpart(d)) for p in procs(d)]])
+Base.sumabs(d::DArray) = sum([fetch(r) for r in [@spawnat p sumabs(localpart(d)) for p in procs(d)]])
+Base.sumabs2(d::DArray) = sum([fetch(r) for r in [@spawnat p sumabs2(localpart(d)) for p in procs(d)]])
+Base.prod(d::DArray) = reduce(Base.MulFun(), d)
+
+Base.any(d::DArray) = reduce(Base.OrFun(), d)
+Base.any(p::Function, d::DArray) = mapreduce(p, Base.OrFun(), d)
```
Comment (amitmurthy) [inline src/DistributedArrays.jl]: This will be slow too. Will change it.

File: src/DistributedArrays.jl [hunk 3]
```diff
@@ -487,4 +487,21 @@ Base.scale!(A::DArray, x::Number) = begin
...
+
+Base.all(d::DArray) = reduce(Base.AndFun(), d)
+Base.all(p::Function, d::DArray) = mapreduce(p, Base.AndFun(), d)
```
Comment (amitmurthy) [inline src/DistributedArrays.jl]: This too.

File: src/DistributedArrays.jl (hunk 2 above)
Comment (jakebolewski) [inline src/DistributedArrays.jl]: Ok, I was just changing them as well.  p should be `Callable`, not `Function`.  You can create an alias to `Base.Callable`

File: src/DistributedArrays.jl (hunk 1 above)
Comment (andreasnoack) [inline src/DistributedArrays.jl]: I think we should fix `mapreduce` and use it for these definitions. We probably just have to define a method similar to `mapreduce` in `reduce.jl` line 156 where the function is substituted with functor.

File: src/DistributedArrays.jl (hunk 2 above)
Comment (amitmurthy) [inline src/DistributedArrays.jl]: using `Base.Callable` is also conflicting with a definition in `reduce.jl` . 

Should be changed along with `mapreduce` in `reduce.jl`

**Pull request merged**